### PR TITLE
Inventory Fixes

### LINKED ISSUE #28: Player inventory broken
```
[Error  : Unity Log] [2022-10-10T17:37:58]..[ERROR] [DEFAULT] [ToolBehaviour] Trying to check if a tool has been unlocked before ToolBehaviour is available. Returning false.
   at ToolBehaviour.IsToolUnlocked(ToolType tool)
   at Travellers.UI.PlayerInventory.HotBarScreen.CheckToolSlot(ToolType type, HotbarSlotType slotType)
   at Travellers.UI.PlayerInventory.HotBarScreen.RefreshToolSlots()
   at Travellers.UI.PlayerInventory.HotBarScreen.RefreshHotbarSlots()
   at Travellers.UI.PlayerInventory.HotBarScreen.ProtectedInit()
   at Travellers.UI.Framework.UIMonoLifetimeController.Initialise()
   at Travellers.UI.Framework.UIObjectFactory.Create(UIElementType elementType, UIFillType fillType, UnityEngine.Transform parent, Boolean isObjectActive, System.String prefabName)
   at Travellers.UI.Framework.HUDState.CreateScreens(System.Collections.Generic.Dictionary`2 screenLookup)
   at Travellers.UI.Framework.WindowState.EnterState()
   at Travellers.UI.Framework.DissociatedStatesStack.PushState(Travellers.UI.Framework.WindowState newState)
   at Travellers.UI.Framework.UIWindowController.PushState(Travellers.UI.Framework.WindowState newState)
   at GameStateMachine.InGameState.OnEnterState()
   at FSM.StateMachine+<SwitchStateWithLoading>c__AnonStorey0.<>m__5(IState o)
   at RSG.Promise`1+<>c__DisplayClass20[[FSM.IState, Assembly-CSharp, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null]].<Then>b__1e(IState v)
   at RSG.Promise`1[[FSM.IState, Assembly-CSharp, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null]].InvokeHandler(System.Action`1 callback, IRejectable rejectable, IState value)
   at RSG.Promise`1[[FSM.IState, Assembly-CSharp, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null]].ActionHandlers(IRejectable resultPromise, System.Action`1 resolveHandler, System.Action`1 rejectHandler)
   at RSG.Promise`1[[FSM.IState, Assembly-CSharp, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null]].Then(System.Action`1 onResolved, System.Action`1 onRejected)
   at RSG.Promise`1[[FSM.IState, Assembly-CSharp, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null]].Then(System.Action`1 onResolved)
   at FSM.StateMachine.SwitchStateWithLoading(IState nextState)
   at FSM.StateMachine.SwitchState(IState nextState)
   at FSM.StateMachine.ProcessState(IState state)
   at FSM.StateMachine.Update(Single deltaTime)
   at GameStateMachine.Bootstrap.Update()```

The `InventoryVisualizer` does not become active so far and thus we get some errors in `HotBarScreen` once its code gets executed. The `OnEnabled()` method of `InventoryVisualizer` does set some stuff that may be needed by the hotbar.
We need to find out why this happens and how we can solve it.

### LINKED ISSUE #39: UI creation error
Right after entering the world for the first time the following error is shown:
image

No error in the server and game server logs:
[server.log](https://github.com/sp00ktober/WorldsAdriftReborn/files/10451295/server.log)
[game-server.log](https://github.com/sp00ktober/WorldsAdriftReborn/files/10451301/game-server.log)

### PR DESCRIPTION
**Pull request by TheSpookyCat**
### Additions/Changes
- `itemData.json` item list file
- ItemHelper for parsing the item list into the consumed formats
- Relocated AuthorityChangeOp usage
- Adds some missing components that were causing problems
- Option to skip the welcome disclaimer on first launch.
- Fixes the "UI Creation Error" seen on startup

### Usage
1. Pull `https://github.com/TheSpookyCat/WorldsAdriftReborn.git`
2. Edit [ItemHelper.cs#L10](https://github.com/TheSpookyCat/WorldsAdriftReborn/blob/new-inventory/WorldsAdriftRebornGameServer/Game/ItemHelper.cs#L10) to use your local `itemData.json` file
3. Launch the game as usual and press tab to access the inventory.

##### Contributors
- IGotzCheese - disclaimer skip & item data

Closes #39
Closes #28

### CODE REVIEW

File: WorldsAdriftRebornGameServer/Game/ItemHelper.cs [hunk 1]
```diff
@@ -0,0 +1,166 @@
+using System.Diagnostics.CodeAnalysis;
+using Bossa.Travellers.Inventory;
+using Improbable.Collections;
+
+namespace WorldsAdriftRebornGameServer.Game
+{
+    public static class ItemHelper
+    {
+        private static Dictionary<string, ValidItem> _allItems = new Dictionary<string, ValidItem>();
+        private const string ItemPath = @"C:\Users\CHANGEME\WorldsAdriftReborn\WorldsAdriftRebornGameServer\Game\itemData.json";
```
Comment (sp00ktober) [inline WorldsAdriftRebornGameServer/Game/ItemHelper.cs:10]: maybe put this file next to the game server .exe (or a folder next to it for potentially more config files) and then load it from a relative path?

File: WorldsAdriftReborn/Patching/LandingScreen/LandingScreenState_Patch.cs [hunk 2]
```diff
@@ -10,12 +9,20 @@ namespace WorldsAdriftReborn.Patching.LandingScreen
     [HarmonyPatch(typeof(LandingScreenState))]
     internal class LandingScreenState_Patch
     {
+        private static void Close() => Application.Quit();
+        private static void SkipYes() => PlayerPrefs.SetInt("skip", 1);
```
[the remark this answers is not in the document]
Comment (mmjr-x) [reply WorldsAdriftReborn/Patching/LandingScreen/LandingScreenState_Patch.cs:13]: Don't think we are using it for anything yet, however it should be quite easy to use. Some good examples can be found here:
https://docs.bepinex.dev/articles/dev_guide/plugin_tutorial/4_configuration.html

File: WorldsAdriftRebornGameServer/Game/Components/ComponentsSerializer.cs [hunk 3]
```diff
@@ -72,17 +73,22 @@ public unsafe static void InitAndSerialize(uint componentId, byte** buffer, uint
...
                     {
                         InventoryState.Data iData = new InventoryState.Data(new InventoryStateData(100,
                                                                                         "{}",
-                                                                                        new Improbable.Collections.List<ScalaSlottedInventoryItem> { },
-                                                                                        new Improbable.Collections.List<ScalaSlottedInventoryItem> { },
-                                                                                        200,
-                                                                                        100,
+                                                                                        ItemHelper.GetDefaultItems(),
+                                                                                        ItemHelper.GetStashItems(true, true),
+                                                                                        10,
```
[the remark this answers is not in the document]
Comment (mmjr-x) [reply WorldsAdriftRebornGameServer/Game/Components/ComponentsSerializer.cs:87]: Could not see that in the PR ;) perhaps width and height should also come from the ItemHelper? ItemHelper.GetDefaultWidth() and ItemHelper.GetDefaultHeight() or something

File: WorldsAdriftRebornGameServer/WorldsAdriftRebornGameServer.csproj [hunk 4]
```diff
@@ -18,6 +18,9 @@
   </ItemGroup>
 
   <ItemGroup>
+    <Reference Include="Assembly-CSharp">
```
[the remark this answers is not in the document]
Comment (mmjr-x) [reply WorldsAdriftRebornGameServer/WorldsAdriftRebornGameServer.csproj:21]: Might be good to remove it then?

File: WorldsAdriftReborn/Patching/LandingScreen/LandingScreenState_Patch.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (sp00ktober) [reply WorldsAdriftReborn/Patching/LandingScreen/LandingScreenState_Patch.cs:13]: we do use it in the mod for some settings :)
https://github.com/sp00ktober/WorldsAdriftReborn/blob/main/WorldsAdriftReborn/Config/ModSettings.cs

**Comment by sp00ktober** [comment]
It seems to partially work, ui error is gone when the game is started in foreground and loading until in game but the inventory is still broken.
however when you put another window in the foreground while loading ingame you will get the ui error popup but have a working inventory (that means you can open it).

Will try to cleanup some server side code after merging to address these issues.

**Review by sp00ktober — APPROVED** [review]

**Pull request merged**